uint32 num1
uint32 num2
uint32 num3
uint32 num4
